File: rust/src/problem02.rs
#[derive(Copy, Clone)]
enum Shape {
    Rock,
    Paper,
    Scissors,
}

impl Shape {
    fn score(&self) -> u32 {
        match self {
            Shape::Rock => 1,
            Shape::Paper => 2,
            Shape::Scissors => 3,
        }
    }

    fn from_char1(c: u8) -> Self {
        match c {
            b'A' => Shape::Rock,
            b'B' => Shape::Paper,
            b'C' => Shape::Scissors,
            _ => panic!("Unknown letter for player 1"),
        }
    }

    fn from_char2(c: u8) -> Self {
        match c {
            b'X' => Shape::Rock,
            b'Y' => Shape::Paper,
            b'Z' => Shape::Scissors,
            _ => panic!("Unknown letter for player 2"),
        }
    }
}


#[derive(Copy, Clone)]
enum Outcome {
    Defeat,
    Draw,
    Win,
}

impl Outcome {
    fn score(&self) -> u32 {
        match self {
            Outcome::Defeat => 0,
            Outcome::Draw => 3,
            Outcome::Win => 6,
        }
    }

    fn from_char(c: u8) -> Self {
        match c {
            b'X' => Outcome::Defeat,
            b'Y' => Outcome::Draw,
            b'Z' => Outcome::Win,
            _ => panic!("Unknown letter for outcome"),
        }
    }
}


fn player2_outcome(player1: Shape, player2: Shape) -> Outcome {
    match (player1, player2) {
        (Shape::Rock, Shape::Rock) => Outcome::Draw,
        (Shape::Rock, Shape::Paper) => Outcome::Win,
        (Shape::Rock, Shape::Scissors) => Outcome::Defeat,
        (Shape::Paper, Shape::Rock) => Outcome::Defeat,
        (Shape::Paper, Shape::Paper) => Outcome::Draw,
        (Shape::Paper, Shape::Scissors) => Outcome::Win,
        (Shape::Scissors, Shape::Rock) => Outcome::Win,
        (Shape::Scissors, Shape::Paper) => Outcome::Defeat,
        (Shape::Scissors, Shape::Scissors) => Outcome::Draw,
    }
}


pub fn solve_a(lines: &mut dyn Iterator<Item = String>) -> u32 {
    let mut result = 0;

    for line in lines {
        let line_as_bytes = line.as_bytes();
        let player1 = Shape::from_char1(line_as_bytes[0]);
        let player2 = Shape::from_char2(line_as_bytes[2]);
        let outcome = player2_outcome(player1, player2);

        result += outcome.score() + player2.score();
    }

    result
}

#[test]
fn solve_a_with_sample_data_returns_15() {
    let sample = indoc::indoc!("
        A Y
        B X
        C Z");
    let mut lines = crate::utils::str_to_iter(sample);

    let actual = solve_a(&mut lines);

    assert_eq!(15, actual);
}


fn player2_shape(player1: Shape, outcome: Outcome) -> Shape {
    match (player1, outcome) {
        (Shape::Rock, Outcome::Defeat) => Shape::Scissors,
        (Shape::Rock, Outcome::Draw) => Shape::Rock,
        (Shape::Rock, Outcome::Win) => Shape::Paper,
        (Shape::Paper, Outcome::Defeat) => Shape::Rock,
        (Shape::Paper, Outcome::Draw) => Shape::Paper,
        (Shape::Paper, Outcome::Win) => Shape::Scissors,
        (Shape::Scissors, Outcome::Defeat) => Shape::Paper,
        (Shape::Scissors, Outcome::Draw) => Shape::Scissors,
        (Shape::Scissors, Outcome::Win) => Shape::Rock,
    }
}

pub fn solve_b(lines: &mut dyn Iterator<Item = String>) -> u32 {
    let mut result = 0;

    for line in lines {
        let line_as_bytes = line.as_bytes();
        let player1 = Shape::from_char1(line_as_bytes[0]);
        let outcome = Outcome::from_char(line_as_bytes[2]);
        let player2 = player2_shape(player1, outcome);

        result += outcome.score() + player2.score();
    }

    result
}

#[test]
fn solve_b_with_sample_data_returns_12() {
    let sample = indoc::indoc!("
        A Y
        B X
        C Z");
    let mut lines = crate::utils::str_to_iter(sample);

    let actual = solve_b(&mut lines);

    assert_eq!(12, actual);
}

File: fsharp/Problem02.fs
module Problem02

type Shape = Rock | Paper | Scissors

type Outcome = Win | Draw | Defeat

let opponent_choice = function
    | 'A' -> Rock
    | 'B' -> Paper
    | 'C' -> Scissors
    | _ -> failwith "Invalid letter for opponent's choice"


let your_choice = function
    | 'X' -> Rock
    | 'Y' -> Paper
    | 'Z' -> Scissors
    | _ -> failwith "Invalid letter for yours choice"
    

let player2_outcome = function
    | (Rock, Rock) -> Draw
    | (Rock, Paper) -> Win
    | (Rock, Scissors) -> Defeat
    | (Paper, Rock) -> Defeat
    | (Paper, Paper) -> Draw
    | (Paper, Scissors) -> Win
    | (Scissors, Rock) -> Win
    | (Scissors, Paper) -> Defeat
    | (Scissors, Scissors) -> Draw


let shape_score = function
    | Rock -> 1
    | Paper -> 2
    | Scissors -> 3
    
    
let outcome_score = function
    | Defeat -> 0
    | Draw -> 3
    | Win -> 6


let solve_a (lines: string seq) =
    lines |> Seq.map (fun line -> (opponent_choice line.[0], your_choice line.[2]))
          |> Seq.map (fun (opponent, your) -> (player2_outcome (opponent, your), your))
          |> Seq.map (fun (outcome, shape) -> outcome_score outcome + shape_score shape)
          |> Seq.sum


let your_outcome = function
    | 'X' -> Defeat
    | 'Y' -> Draw
    | 'Z' -> Win
    | _ -> failwith "Invalid letter for yours outcome"


let player2_shape = function
    | (Rock, Defeat) -> Scissors
    | (Rock, Draw) -> Rock
    | (Rock, Win) -> Paper
    | (Paper, Defeat) -> Rock
    | (Paper, Draw) -> Paper
    | (Paper, Win) -> Scissors
    | (Scissors, Defeat) -> Paper
    | (Scissors, Draw) -> Scissors
    | (Scissors, Win) -> Rock


let solve_b (lines: string seq) =
    lines |> Seq.map (fun line -> (opponent_choice line.[0], your_outcome line.[2]))
          |> Seq.map (fun (opponent, outcome) -> (outcome, player2_shape (opponent, outcome)))
          |> Seq.map (fun (outcome, shape) -> outcome_score outcome + shape_score shape)
          |> Seq.sum
